beps/0012-metrics-service/README.md

```diff
@@ -0,0 +1,351 @@
+
+This approach leverages industry standards while focusing the `MetricsService` on distinct Backstage concerns, following the same pattern as other core services (DatabaseService builds on Knex, LoggerService builds on Winston, HttpRouterService builds on Express, etc.).
+
+## Motivation
+
+There is no guidance when it comes to metrics instrumentation. While individual plugins may implement their own metrics, there's no standardized approach for metrics collection, naming, or configuration.
+
+By providing a core metrics service:
+
+- **Plugin Authors** and the **Community** gain a straightforward way to address metrics instrumentation and can focus on business logic instead of needing to reimplement metrics plumbing.
+- **Backstage Admins** receive a reliable stream of metrics from the core system for monitoring, alerting, and troubleshooting.
+
+### Goals
+
+- Plugin-scoped metric namespacing
+- Consistent metrics patterns across all plugins
+- Aligned with OpenTelemetry industry standards
+- Provide a familiar interface as other core services
+- Standardized initialization and lifecycle management
+
+The catalog and scaffolder plugins will be updated to use the new metrics service in the initial alpha release.
+
+### Non-Goals
+
+- Adding metrics to plugins that don't currently have it (outside of catalog and scaffolder)
+- Tracing and other telemetry concerns are out of scope for this BEP.
+- Refactoring the existing `LoggerService`. Future work to unify observability related concerns would be ideal, but not a goal.
+
+## Proposal
+
+Following similar patterns to other core services, create a new `RootMetricsService` responsible for initializing the OpenTelemetry SDK, root-level concerns, and the creation of plugin-specific `MetricsService` instances. The root service delegates to the plugin-scoped `MetricsService` to initialize a meter for each registered plugin based on the `service.name` and optional `service.version` provided in the app's `app-config.yaml` file. This is based on the recommendation in the OpenTelemetry [documentation](https://opentelemetry.io/docs/languages/js/instrumentation/#acquiring-a-meter).
+
+## Naming Conventions
+
+All Backstage metrics follow this hierarchical pattern:
+
+`backstage.{scope}.{scope_name}.{metric_name}`
+
+**Where:**
+
+- `backstage` is the root namespace for all Backstage metrics
+- `{scope}` is the system scope (either **plugin** or **core**)
+- `{scope_name}` is the name of the plugin or core service (e.g., `catalog`, `scaffolder`, `database`)
+- `{metric_name}` is the hierarchical metric name as provided by the plugin author (e.g., `entities.processed.total`, `tasks.completed.total`)
+
+### Scope
+
+The `scope` represents where it belongs in the Backstage ecosystem.
+
+- `plugin` - A plugin-specific metric (e.g. `backstage.plugin.catalog.entities.count`)
+- `core` - A core service metric (e.g. `backstage.core.database.connections.active`)
```

Comment:
Suggestion: `service` instead of `core`. Seeing a `backstage.service.scheduler.` metric makes sense to me.

What do we do with plugin scoped services vs root scoped services? Will we standardize on ALL services being under `backstage.service.foo` but plugin-scoped ones have an automatically injected `pluginId` attribute? Or is it up to the plugin scoped services themselves to add such an attribute with that exact name as a standard?

Reply:
My original intention was to separate something occuring within the framework vs. a plugin. So maybe, we don't represent services under _core_, but we still keep it and the service name comes next within the hierarchy. 

```ts
backstage.plugin.database.connections.active
backstage.core.database.connections.active
```

or maybe leverage `framework` or `root`

```ts
backstage.plugin.database.connections.active
backstage.{framework|root}.database.connections.active
```

Reply:
I went down the _services vs plugins_ path a little too much and think that it was the incorrect approach. Here is a table I put together of what the existing metrics would look like if we were to separate `framework vs plugin` scope. The end goal is the framework provides native instrumentation across all core services. So, end users get native telemetry there for free and only need to focus on business concerns within the confinements of their plugin. 

Using the scheduler service as an example, regardless of which plugin schedules something, they should all get some level of instrumentation provided by the framework. 

<img width="457" height="445" alt="image" src="https://github.com/user-attachments/assets/6987b806-845a-47fd-a795-1192a8b5f843" />

Reply:
FYI, I don't think that works either. It's better, but we need a `plugin` and/or `service` namespace in there. I have reached out to the OTEL community for feedback because naming is hard and there is probably someone out there that enjoys this more than I do.

Reply:
Let's try something different. I've created a [PR](https://github.com/backstage/backstage/pull/30845) that focuses on providing documentation around naming conventions for metrics. This will help focus full attention on this topic and it's something that can be published without ever implementing a metrics service.